src/Core/Batch/BatchTrait.php

```diff
@@ -0,0 +1,195 @@
+                    memory_get_usage()
+                )
+            );
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns an array representation of a callback which will be used to write
+     * batch items.
+     *
+     * @return array
+     */
+    protected abstract function getCallback();
+
+    /**
+     * @param array $options [optional] {
+     *     Configuration options.
+     *
+     *     @type resource $debugOutputResource A resource to output debug output
+     *           to.
+     *     @type bool $debugOutput Whether or not to output debug information.
+     *           **Defaults to** false.
+     *     @type array $batchOptions A set of options for a BatchJob.
+     *           {@see \Google\Cloud\Core\Batch\BatchJob::__construct()} for
+     *           more details.
+     *           **Defaults to** ['batchSize' => 1000,
+     *                            'callPeriod' => 2.0,
+     *                            'workerNum' => 2].
+     *     @type array $clientConfig A config used to construct the client upon
+     *           which requests will be made.
+     *     @type BatchRunner $batchRunner A BatchRunner object. Mainly used for
+     *           the tests to inject a mock. **Defaults to** a newly created
+     *           BatchRunner.
+     *     @type string $identifier An identifier for the batch job.
+     *     @type string $batchMethod The name of the batch method used to
+     *           deliver items.
+     * }
+     * @throws \InvalidArgumentException
+     */
+    private function setCommonBatchProperties(array $options)
+    {
+        if (!isset($options['identifier'])) {
+            throw new \InvalidArgumentException(
+                'A valid identifier is required in order to register a job.'
+            );
+        }
+
+        if (!isset($options['batchMethod'])) {
+            throw new \InvalidArgumentException(
+                'A batchMethod is required.'
+            );
+        }
+
+        $this->batchMethod = $options['batchMethod'];
+        $this->identifier = $options['identifier'];
+        $this->debugOutputResource = isset($options['debugOutputResource'])
+            ? $options['debugOutputResource']
+            : null;
+        $this->debugOutput = isset($options['debugOutput'])
+            ? $options['debugOutput']
+            : false;
+        $this->clientConfig = isset($options['clientConfig'])
+            ? $options['clientConfig']
+            : [];
+        $batchOptions = isset($options['batchOptions'])
+            ? $options['batchOptions']
+            : [];
+        $this->batchOptions = $batchOptions + [
+            'batchSize' => 1000,
+            'callPeriod' => 2.0,
+            'workerNum' => 2
+        ];
+        $this->batchRunner = isset($options['batchRunner'])
+            ? $options['batchRunner']
+            : new BatchRunner();
+        $this->batchRunner->registerJob(
+            $this->identifier,
+            [$this, 'send'],
```

Comment:
Having the common `send` function is good for maintenance, but additional function call adds certain cost within the infinite loop in the daemon.

Does it make sense to directly register the function, and remove the `send` and `getCallback` functions? There will be some copy-paste code here and there, but if it gains performance, it's gonna be a good maintenance tax to pay for the better performance.